### PR TITLE
[libobject] Common metadata for declared objects

### PR DESCRIPTION
**Pull request by ejgallego**
This is a first implementation of CEP https://github.com/coq/ceps/pull/65

We add a common metadata object to atomic libobjects, so we can store
documentation, location, names, and other interesting attributes that
are needed by clients to implement functionality.

As we allow storing names of the objects, we can index them by name to
provide efficient reverse lookup.

We export an experimental API for clients to consume this metadata.

PR https://github.com/coq/coq/pull/16261 is implemented on top of this one.

### CODE REVIEW

File: library/libobject.mli [hunk 1]
```diff
@@ -72,46 +72,54 @@ type substitutivity = Dispose | Substitute | Keep | Anticipate
...
-}
+  type t
+
+  val unfiltered : t
+
+  val make : finite:bool -> string CAst.t list -> t
+  (** Anomaly when the list is empty. *)
 
-val unfiltered : open_filter
+  module Category : sig
```
Comment (SkySkimmer) [inline library/libobject.mli:84]: Not sure this should be a submodule of Open_filter

File: library/libobject.mli [hunk 2]
```diff
@@ -72,46 +72,54 @@ type substitutivity = Dispose | Substitute | Keep | Anticipate
...
 
-val simple_open : ?cat:category -> ('i -> 'a -> unit) ->
-  open_filter -> 'i -> 'a -> unit
-(** Combinator for making objects with simple category-based open
-   behaviour. When [cat:None], can be opened by Unfiltered, but also
-   by Filtered with a negative set. *)
 
-val filter_and : open_filter -> open_filter -> open_filter option
-(** Returns [None] when the intersection is empty. *)
+  val and_ : t -> t -> t option
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply library/libobject.mli:106]: Ok.

[the remark this answers is not in the document]
Comment (ejgallego) [reply library/libobject.mli:106]: Actually OCaml doesn't allow that.

[the remark this answers is not in the document]
Comment (SkySkimmer) [reply library/libobject.mli:106]: Ah I tried with `or` but not `and`

File: library/libobject.mli (hunk 1 above)
Comment (ejgallego) [reply library/libobject.mli:84]: I agree. Should category go into `Data.t` ? I think we can do a bit better with this API in general, but I dunno.

Comment (SkySkimmer) [reply library/libobject.mli:84]: I would put it as Libobject.Category
No need to pile on inner modules

Comment (ejgallego) [reply library/libobject.mli:84]: Actually my motivation to put it into `Open_filter` is that actually `Open_filter` has access to the private implementation of `Category.t`. So I'm not sure what the best thing is here.

File: library/libobject.mli [hunk 3]
```diff
@@ -132,24 +140,53 @@ val ident_subst_function : substitution * 'a -> 'a
    will hand back two functions, the "injection" and "projection"
    functions for dynamically typed library-objects. *)
 
+(** Common metadata for objects that are stored in modules / .vo files *)
+module Data : sig
+
+  (* EJGA: Another apporach would be to make this monomorphic and store
+     just a pair, but I've opted for the "strongest" API for now *)
+  type 'a t
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply library/libobject.mli:149]: As it stands today yes, I've opted for the constructor/projector setup tho as in general is more flexible if we need to change the semantics. OMMV.

File: library/libobject.mli (hunk 1 above)
Comment (ejgallego) [reply library/libobject.mli]: > I would put it as Libobject.Category No need to pile on inner modules

I mean if we should store the category in the obj itself instead of encoding the filtering logic in the closure.

For the original motivation of putting it inside `Open_filter` see my comment about internals.

Comment (SkySkimmer) [reply library/libobject.mli]: We could imagine categories getting more complicated than 1 string, eg like in https://github.com/coq/coq/pull/14892#issuecomment-922903956
Not sure how much we should consider these dreams of the future while writing code today though

File: library/libobject.mli (hunk 2 above)
[the remark this answers is not in the document]
Comment (Alizter) [reply library/libobject.mli:107]: What about `(&)`?

File: library/libobject.mli (hunk 3 above)
[the remark this answers is not in the document]
Comment (ejgallego) [reply library/libobject.mli]: I've pushed a commit indeed to use a tuple for now, and got rid of the copy and obj.

File: interp/abbreviation.ml [hunk 4]
```diff
@@ -73,11 +73,11 @@ let classify_abbreviation (local,_) =
 
 let inAbbreviation : Id.t -> (bool * abbreviation) -> obj =
   declare_named_object {(default_object "ABBREVIATION") with
-    cache_function = cache_abbreviation;
-    load_function = load_abbreviation;
-    open_function = simple_open open_abbreviation;
-    subst_function = subst_abbreviation;
-    classify_function = classify_abbreviation }
+    cache_function = cache_abbreviation
+    ; load_function = load_abbreviation
```
Comment (SkySkimmer) [inline interp/abbreviation.ml:77]: This is ugly and also unnecessary diff

File: library/lib.ml [hunk 5]
```diff
@@ -168,14 +168,17 @@ let add_leaf_entry leaf =
...
   lib_state := { !lib_state with lib_stk }
 
-let add_discharged_leaf obj =
+let add_discharged_leaf (data, obj) =
   let newobj = Libobject.rebuild_object obj in
-  Libobject.cache_object (prefix(),newobj);
-  add_leaf_entry (AtomicObject newobj)
+  Libobject.cache_object (prefix(), newobj);
+  (* XXX fixme *)
+  let data = Libobject.Data.empty in
```
Comment (SkySkimmer) [inline library/lib.ml:175]: What's going on here?

**Comment by SkySkimmer** [comment]
coqchk is broken (needs values.ml update)

It would be nice to clean up spurious diffs (eg ssrparser)

I don't think I really understand the value of this system but I'm willing to merge.

File: interp/abbreviation.ml (hunk 4 above)
Comment (ejgallego) [reply interp/abbreviation.ml:77]: Yes, it is an artifact of previous versions that did update this field tho.

**Comment by ejgallego** [comment]
> coqchk is broken (needs values.ml update)
> 
> It would be nice to clean up spurious diffs (eg ssrparser)

Yes to both.

> I don't think I really understand the value of this system but I'm willing to merge.

Would you like to learn more about what I'm trying to do (in a call for example?)

**Comment by gares** [comment]
What I don't get is why the extra arguments (eg the loc) are not arguments of the function returned by libobject. It would make the diff much smaller, for example.

**Comment by ejgallego** [comment]
@coqbot request full CI

**Comment by ejgallego** [comment]
@coqbot run full ci

**Comment by ejgallego** [comment]
Ok, I've cleaned up this and the dependency #16261 ; I think I'm happy with the current iteration of both PRs and IMHO that would be great functionality to have in 8.17

The main FAQ based on the discussion is why do we make the meta-data common instead of sealing it? There are three main reasons that I can think of:
- sealing the data requires a way to get it out, this is more complex and usually limited to a single datatype. For example we would need N indexes in the summary to support location for N objects, here we can support `GlobRef` and Notation as a first step just fine,
- it seems wasteful to duplicate a map that is already there, moreover we need to sync them, and the objects are already tracked,
- the current API is backwards compatible and pretty easy, we can add for example a PR for doc attributes with a few lines.

But this doesn't bar us IMHO from updating the API in the future.

**Comment by SkySkimmer** [comment]
>Ok, I've cleaned up this and the dependency #16261 ; I think I'm happy with the current iteration of both PRs and IMHO that would be great functionality to have in 8.17

IMO the discussion in the coq call did not conclude and was cut by time constraints, with the branch imminent this is unlikely to get in 8.17.

**Comment by ejgallego** [comment]
@coqbot run full ci

**Comment by ejgallego** [comment]
> IMO the discussion in the coq call did not conclude and was cut by time constraints, with the branch imminent this is unlikely to get in 8.17.

How do you propose to proceed?

My reasoning is that the location feature is worth trying for 8.17 (having locations really changes the game for some use cases) but I don't know anymore.

**Comment by ejgallego** [comment]
@coqbot run full ci

**Comment by Zimmi48** [comment]
The PR now conflicts with changes merged in master and targeting 8.18, making it completely unrealistic to consider this PR for inclusion in 8.17.

**Comment by ejgallego** [comment]
Not an urgency for 8.18.
